File: LeetCode/Java/src/solutions/p0129_SumRoottoLeafNumbers/P0129.java
package solutions.p0129_SumRoottoLeafNumbers;

import auxiliary.TreeNode;

public class P0129 {
    public int sumNumbers(TreeNode root) {
        /*
         * Solution v1: Recursive DFS
         *
         * Runtime: 0 ms @ (beats) 100%
         * Memory Usage: 40.4 MB @ (beats) 28.58%
         */

        return dfs(root, 0);
    }

    public static int dfs(TreeNode root, int val) {
        if (root == null) {
            return 0;
        }

        int new_v = val * 10 + root.val;

        if (root.left == null && root.right == null) {
            return new_v;
        }

        return dfs(root.left, new_v) + dfs(root.right, new_v);
    }
}

File: LeetCode/Python/solutions/p0129.py
import collections
from typing import Optional

from Python.auxiliary.TreeNode import TreeNode


class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        #
        # Solution v1: BFS
        #
        # Runtime: 30 ms @ (beats) 81.13%
        # Memory Usage: 13.9 MB @ (beats) 54.57%
        #
        # ans = []
        #
        # q = collections.deque()
        # q.append(("", root))
        #
        # while q:
        #     num, curr = q.popleft()
        #
        #     if not curr:
        #         continue
        #
        #     num = f"{num}{curr.val}"
        #
        #     if curr.left or curr.right:
        #         q.append((num, curr.left))
        #         q.append((num, curr.right))
        #     else:
        #         ans.append(int(num))
        #
        # return sum(ans)

        #
        # Solution v1: BFS
        # int instead of str
        #
        # Runtime: 37 ms @ (beats) 33.24%
        # Memory Usage: 13.9 MB @ (beats) 54.57%
        #
        # ans = []
        #
        # q = collections.deque()
        # q.append((0, root))
        #
        # while q:
        #     num, curr = q.popleft()
        #
        #     if not curr:
        #         continue
        #
        #     num = num * 10 + curr.val
        #
        #     if curr.left or curr.right:
        #         q.append((num, curr.left))
        #         q.append((num, curr.right))
        #     else:
        #         ans.append(num)
        #
        # return sum(ans)

        #
        # Solution v2: BFS Recursive
        #
        # Runtime: 32 ms @ (beats) 70.48%
        # Memory Usage: 13.9 MB @ (beats) 54.57%
        #

        def dfs(head, val):
            if not head:
                return 0

            new_v = val * 10 + head.val

            if not head.left and not head.right:
                return new_v

            return dfs(head.left, new_v) + dfs(head.right, new_v)

        return dfs(root, 0)
